rfc/rfc-77/rfc-77.md

```diff
@@ -0,0 +1,323 @@
+support needs to be added to merge records from MDT partition hosting SI (with generalized approach that can be used for
+any future non-unique record key based dataset).
+
+The data structure built for holding the merged log records should enable to search similar keys efficiently with 
+following consideration:
+1. Should allow spilling to disk (to prevent OOM). A pure in-memory data structure will not be production ready, but 
+can be used only for POC
+2. Should allow for searching (and retrieving) based on two keys as efficiently as possible. First search will be based 
+on `secondary-key` and second search will be based on `record-key`. Hence, a single level Map used by 
+`HoodieMergedLogRecordScanner` will not work.
+3. Should allow for efficient removal of records (deleted record need to be removed). Hence, any data structure that 
+uses a flat array will not be efficient. 
+4. Should allow for efficient insertion of records (for inserting merged record and for buffering fresh records). 
+
+The [initial POC](https://github.com/apache/hudi/pull/10625) makes use of an in-memory nested maps - with the first level keyed by `secondary-key`  
+and the second level keyed by`record-key`. However, the final design should allow spilling to disk.
+
+Considering the above requirements, the proposal is to extend `HoodieKey` to support secondary keys. We will make `HoodieKey` generic and create a subclass `HoodieSecondaryKey` where the key can be a generic type.
+```java
+public class HoodieKey<T extends Serializable> implements Serializable {
+
+  private T recordKey;
+  private String partitionPath;
+
+  // Required for serializer
+  public HoodieKey() {}
+
+  public HoodieKey(T recordKey, String partitionPath) {
+    this.recordKey = recordKey;
+    this.partitionPath = partitionPath;
+  }
+
+  // Getters and setters
+  public T getRecordKey() {
+    return recordKey;
+  }
+
+  public void setRecordKey(T recordKey) {
+    this.recordKey = recordKey;
+  }
+
+  public String getPartitionPath() {
+    return partitionPath;
+  }
+
+  public void setPartitionPath(String partitionPath) {
+    this.partitionPath = partitionPath;
+  }
+}
+```
+
+Since `HoodieKey` is now generic, `HoodieSecondaryKey` can specify any type for its key by setting the type parameter
+when it extends `HoodieKey`.
+
+```java
+public class HoodieSecondaryKey<K extends Serializable> extends HoodieKey<K> {
+
+  // Additional properties and methods specific to HoodieSecondaryKey can be added here
+
+  public HoodieSecondaryKey(K recordKey, String partitionPath) {
+    super(recordKey, partitionPath);
+  }
+
+  // Example method that might use the generic type K
+  public void someMethod() {
+    K key = getRecordKey();
+    // Process key which is of type K
+  }
+}
+```
+
+The primary advantage of this approach is that the merge algorithm does not change. We interpret the key in the scanner
```

Comment:
lets not leak secondary index/key concepts (which are 2 layers higher than merge handle or base/log merging) into these layers. code/comments/docs should stay at the level of merge keys.